Fix #18782 (Add streaming logic in ChatHuggingFace)

lunara-kim opened the pull request:
- Add functions (_stream, _astream)
- Connect to _generate and _agenerate

Thank you for contributing to LangChain!

- [x] **PR title**: "community: Add streaming logic in ChatHuggingFace"

- [x] **PR message**: ***Delete this entire checklist*** and replace with
    - **Description:** Addition functions (_stream, _astream) and connection to _generate and _agenerate
    - **Issue:** #18782
    - **Dependencies:** none
    - **Twitter handle:** @lunara_x

lunara-kim committed:
Fix #18782 (Add streaming logic in ChatHuggingFace)

- Add functions (_stream, _astream)
- Connect to _generate and _agenerate

eyurtsev commented on libs/community/langchain_community/chat_models/huggingface.py:
```diff
@@ -71,13 +80,50 @@ def validate_llm(cls, values: dict) -> dict:
         run_manager: Optional[CallbackManagerForLLMRun] = None,
         **kwargs: Any,
     ) -> ChatResult:
+        if self.streaming:
```
we generally don't want to support self`.streaming ` on _generate going forward -- we want users to use .stream() or .astream() for streaming exclusively

eunhye1kim replied:
@eyurtsev 
I've updated the code based on lint recommendations. 😊
However, the discussion about removing self.streaming=True might need further consideration. 😢

Notably, ChatOpenAI supports streaming=True here:
(https://github.com/langchain-ai/langchain/blob/53ac1ebbbccbc16ce57badf08522c6e59256fdfe/libs/community/langchain_community/chat_models/openai.py#L150)

As for LLMChain, it currently requires the addition of streaming support:
(https://github.com/langchain-ai/langchain/blob/53ac1ebbbccbc16ce57badf08522c6e59256fdfe/libs/langchain/langchain/chains/llm.py#L37)

I'll be looking forward to any updates on this topic and am ready to follow up accordingly.

eyurtsev reviewed (commented):
Looks reasonable, do you think you'd be able to lint the PR?

baskaryan commented:
cc @aymeric-roucher

lunara-kim committed:
Apply lint fixes and _stream fix to ChatHuggingFace

- Resolve E501 line too long warnings
- Remove unused imports
- Clean up and sort imports with ruff
- Modified to call llm.stream instead of llm._stream
- Solved the problem of an error occurring due to streaming variable at lint (python 3.11)

baskaryan commented on libs/community/langchain_community/chat_models/huggingface.py:
```diff
@@ -71,6 +78,37 @@ def validate_llm(cls, values: dict) -> dict:
+    ) -> Iterator[ChatGenerationChunk]:
+        request = self._to_chat_prompt(messages)
+
+        for data in self.llm._stream(request, **kwargs):
```
should this just call self.llm.stream?

eunhye1kim replied:
@baskaryan 

Yes, this is a wrapper for the HuggingFaceTextGenInference, HuggingFace Endpoint, and HuggingFace_Hub models.
So I checked that it is working properly by calling _stream of llm.
For reference, I tested as follows.

(in opengpts.. backend/app/llms.py)

@lru_cache(maxsize=1)
def get_tgi_llm():
    huggingface_hub.login(os.getenv("HUGGINGFACE_TOKEN"))
    llm = HuggingFaceTextGenInference(
        inference_server_url="http://myinferenceserver/",
        max_new_tokens=1024,
        top_k=10,
        top_p=0.95,
        typical_p=0.95,
        temperature=0.3,
        repetition_penalty=1.1,
        streaming=True,
        callbacks=[],
        server_kwargs={
            "headers": {
                "Content-Type": "application/json",
            }
        },
    )
    return ChatHuggingFace(llm=llm, model_id="HuggingFaceH4/zephyr-7b-beta") # setting a model_id for initializing the tokenizer (the another bug)


(in opengpts.. backend/app/agent.py)
def get_agent_executor(
    ...)
    if agent == AgentType.TGI:
        llm = get_gauss_llm()
        system_message = my_prompt
        return get_gauss_agent_executor(
            tools, llm, system_message, interrupt_before_action, CHECKPOINTER
        )
    else:
        raise ValueError("Unexpected agent type")

eunhye1kim replied:
@baskaryan 

After reading the comment you left, I checked not only the class I used, but also the huggingface endpoint and huggingfacehub.
The HuggingFaceEndpoint class had stream-related content, but HuggingFaceHub class only had generate logic without any such streaming logic.
It seems like I need to explicitly specify the type of llm at ChatHuggingFace as Union type in the initialization step, 
but I want to respect the history of changing the type to Any instead of Union.
Instead, I will add streaming support code to the HuggingFaceHub class and check 
whether it works properly and whether it can be integrated with ChatHuggingFace, and then I'll upload the modified code.

baskaryan replied:
@eunhye1kim just mean should we use `self.llm.stream` instead of `self.llm._stream`

eunhye1kim replied:
@baskaryan I agree. So I changed the code. In the case of HuggingFaceHub, I think streaming support logic may not be necessary because it is a deprecated class.
